[#2291, #5770] Tie item uses recovery into calendar system

arbron opened the pull request:
Allows recovery for for periods of "day", "dawn", or "dusk" to be triggered by the passage of world time rather that through the rest dialog. This is gated by a setting that by default is tied to whether the calendar is enabled, but can be set to manual if needed (for example, if using a module-provided calendar).

Some changes were made to the `recoverUses` API to support the passage of multiple relevant recovery periods if the time was advanced more than one day. This also fixes the bug with short rests when using the gritty realism rest setting.

When a time passes the system checks to see if any relevant recovery periods should trigger and if so, triggers recovery for all world and actor items. If any changes were made, it then creates a chat message to indicate the time passage and display any deltas. These deltas are gated by observer permissions.

<img width="312" height="312" alt="Time Passage Messages" src="https://github.com/user-attachments/assets/b526f9cc-0ac7-4603-bc27-e01d6f6f2e06" />

The created chat message also has a description of the time that passed (e.g. "The sun has risen." or "Two days have passed."). In order to support better localization of these times, some updates have been made to the number formatting system to support the `words` (and `numerals`) settings when formatting with units.

One potential point of improvement is how we handle selection of recovery periods. At the moment, we only select the first recovery profile that matches a provided period. This may cause issues for cases where a "dawn" and "dusk" may have passed in one update. If you had an item that was only usable during the daytime, you might say it recovers all uses at dawn and loses them at dusk. This works fine so long as you do not advance past both periods at the same time. In the future the recovery uses method might be update with the option of selecting multiple profiles so long as they are not marked as mutually exclusive.

Closes #2291
Closes #5770

arbron committed:
[#2291, #5770] Tie item uses recovery into calendar system

Allows recovery for for periods of "day", "dawn", or "dusk" to be
triggered by the passage of world time rather that through the
rest dialog. This is gated by a setting that by default is tied to
whether the calendar is enabled, but can be set to manual if
needed (for example, if using a module-provided calendar).

Some changes were made to the `recoverUses` API to support the
passage of multiple relevant recovery periods if the time was
advanced more than one day. This also fixes the bug with short
rests when using the gritty realism rest setting.

When a time passes the system checks to see if any relevant
recovery periods should trigger and if so, triggers recovery for
all world and actor items. If any changes were made, it then
creates a chat message to indicate the time passage and display
any deltas. These deltas are gated by observer permissions.

The created chat message also has a description of the time that
passed (e.g. "The sun has risen." or "Two days have passed."). In
order to support better localization of these times, some updates
have been made to the number formatting system to support the
`words` (and `numerals`) settings when formatting with units.

One potential point of improvement is how we handle selection of
recovery periods. At the moment, we only select the first recovery
profile that matches a provided period. This may cause issues for
cases where a "dawn" and "dusk" may have passed in one update. If
you had an item that was only usable during the daytime, you might
say it recovers all uses at dawn and loses them at dusk. This works
fine so long as you do not advance past both periods at the same
time. In the future the recovery uses method might be update with
the option of selecting multiple profiles so long as they are not
marked as mutually exclusive.

Closes #2291
Closes #5770

arbron committed:
[#2291, #5770] Clean up type definitions

Fyorl commented on lang/en.json:
```diff
@@ -1358,6 +1377,17 @@
+    "Midnight": "A new day has begun.",
+    "Sunrise": "The sun has risen.",
+    "Sunset": "The sun has set.",
+    "SunsetSunrise": "The sun has set and a new dawn of a new day has come.",
```
```suggestion
    "SunsetSunrise": "The sun has set and the dawn of a new day has come.",
```

Fyorl commented on module/data/calendar/calendar-data.mjs:
```diff
@@ -327,6 +328,122 @@ export default class CalendarData5e extends foundry.data.CalendarData {
+      };
+
+      const itemUpdates = [];
+      await performItemUpdates(game.items, itemUpdates);
```
An item not on an Actor I would imagine is kind of 'inert', no? When would we expect that such an item recovers its uses over the passage of time?

Fyorl commented on module/data/calendar/calendar-data.mjs:
```diff
@@ -327,6 +328,122 @@ export default class CalendarData5e extends foundry.data.CalendarData {
+      if ( timePassageData[d] ) periods.set(p, timePassageData[d]);
+    }
+
+    const deltas = [];
```
Avoid sweeping `game.items` and `game.actors` if there's nothing to do.
```suggestion
    if ( !periods.size ) return;
    const deltas = [];
```

Fyorl commented on module/data/calendar/calendar-data.mjs:
```diff
@@ -327,6 +328,122 @@ export default class CalendarData5e extends foundry.data.CalendarData {
+   * @param {TimePassageData} timePassageData
+   */
+  static async handleTimePassage(timePassageData) {
+    const periods = new Map();
```
Seems we only create a message if we have `deltas` and we only have `deltas` if we are doing calendar recovery.

```suggestion
    if ( !dnd5e.settings.calendarConfig.manualRecovery ) return;
    const periods = new Map();
```

Then we don't need the if-statement

arbron replied:
My intention here is to give modules a chance to use the message even if nothing is recovered. In that case the `dnd5e.preCreateTimePassedMessage` hook is fired and modules can set `create: true` and make any changes to the message they like.

Fyorl replied:
Yeah, we can still call the hook. Maybe it's a bit neater to check `manualRecovery` and `periods.size` near the top here and call the hook before early-return.

Fyorl commented on module/data/calendar/calendar-data.mjs:
```diff
@@ -327,6 +328,122 @@ export default class CalendarData5e extends foundry.data.CalendarData {
+   * Trigger recovery for any world items based on the passage of time.
+   * @param {TimePassageData} timePassageData
+   */
+  static async handleTimePassage(timePassageData) {
```
I think so long as we bail early if `!periods.size` then scanning all items on all world actors is probably fine, since we would only ever do it if there's a long rest or the DM is advancing time a lot.

But even a normal adventure will import a fair few actors. I had some considerations for how we could improve performance so as to not have to worry about this method at all:
1. Only scan `character`s, `vehicle`s, and `npc`s that are in `group`s or flagged as `important`.
2. Move the `performItemUpdates` closure out to a static helper.
3. `getRollData` is pretty expensive, but we call it indiscriminately even though it's only necessary for formula-type recovery. Could we make the `rollData` param optional and have the item call `getRollData` on itself only when absolutely required?

arbron replied:
I've eliminated world item updates so I can inline the `performItemUpdates` helper and I've added on-demand `getRollData`. I've also moved the updates themselves into a batch update.

Fyorl commented on module/data/calendar/calendar-data.mjs:
```diff
@@ -327,6 +328,122 @@ export default class CalendarData5e extends foundry.data.CalendarData {
+      const { value, unit } = convertTime(timePassageData.midnights, "day", { strict: false });
+      const number = formatTime(value, unit, { words: true });
+      message = game.i18n.format(`DND5E.CALENDAR.TimePassage.TimePassed.${pr.select(value)}`, {
+        number: number.capitalize(), numberLc: number
```
What's `numberLc` for?

arbron replied:
So languages that don't put the number at the start of the sentence can control capitalization.

Fyorl commented on module/data/chat-message/_types.mjs:
```diff
@@ -60,6 +60,19 @@
+/**
+ * @typedef DocumentDeltasData
+ * @param {IndividualDeltaData} deltas  Data deltas for an document update.
+ * @param {string} uuid                 UUID of the document to which the deltas apply.
```
```suggestion
 * @property {IndividualDeltaData[]} deltas  Data deltas for a document update.
 * @property {string} uuid                 UUID of the document to which the deltas apply.
```

Fyorl commented on module/data/settings/calendar-setting.mjs:
```diff
@@ -19,9 +19,26 @@ export class CalendarConfigSetting extends foundry.abstract.DataModel {
+   * @type {boolean}
+   */
+  get manualRecovery() {
+    return this.dailyRecovery === "manual" || (!this.dailyRecovery && !this.enabled);
```
```suggestion
    return (this.dailyRecovery === "manual") || (!this.dailyRecovery && !this.enabled);
```

Fyorl commented on module/data/calendar/calendar-data.mjs:
```diff
@@ -327,6 +328,122 @@ export default class CalendarData5e extends foundry.data.CalendarData {
+   */
+  static onTimePassage(worldTime, deltaTime, options, userId) {
+    if ( !game.user.isActiveGM || (deltaTime <= 0) ) return;
+    const timePassageData = { worldTime, deltaTime, ...options.dnd5e.deltas };
```
```suggestion
    const timePassageData = { worldTime, deltaTime, ...(options.dnd5e?.deltas ?? {}) };
```

Fyorl commented on module/data/calendar/calendar-data.mjs:
```diff
@@ -327,6 +328,122 @@ export default class CalendarData5e extends foundry.data.CalendarData {
+          const result = await item.system.recoverUses?.(periods, item.getRollData()) ?? {};
+          if ( !foundry.utils.isEmpty(result?.updates) ) {
+            deltas.push({ uuid: item.uuid, deltas: IndividualDeltaField.getDeltas(item, result.updates) });
+            updates.push({ _id: item.id, ...result.updates });
```
Do we need to handle `result.destroy`?

arbron replied:
Yep, another feature that is newer than this PR :)

Fyorl commented on module/data/chat-message/time-passed-message-data.mjs:
```diff
@@ -0,0 +1,64 @@
+import ChatMessageDataModel from "../abstract/chat-message-data-model.mjs";
+import ActivationsField from "./fields/activations-field.mjs";
```
Not used

arbron committed:
[#2291] Don't hit world items, batch updates, handle deletes

- Only check items on actors, no standalone world items
- Handle destroyed consumables
- Avoid recovery if no periods set
- Only fetch roll data if item has a formula recovery
- Batch recovery updates

arbron committed:
[#2291] Swap to checking cached setting

Fyorl commented on module/data/calendar/calendar-data.mjs:
```diff
@@ -327,6 +328,121 @@ export default class CalendarData5e extends foundry.data.CalendarData {
+  /* -------------------------------------------- */
+
+  /**
+   * Trigger recovery for any world items based on the passage of time.
```
```suggestion
   * Trigger recovery for any world actors based on the passage of time.
```

Fyorl commented on module/data/item/templates/activities.mjs:
```diff
@@ -349,6 +358,9 @@ export default class ActivitiesTemplate extends SystemDataModel {
     };
 
+    if ( !rollData && (this.activities.some(a => a.uses?.recovery.some(u => u.type === "formula"))
+      || this.uses.recovery.some(u => u.type === "formula")) ) rollData = this.parent.getRollData();
```
I think we can make this a teeny bit faster.

```suggestion
    if ( !rollData && this.uses.recovery.some(u => u.type === "formula")) ) rollData = this.parent.getRollData();
```

Then in the `for` loop below:
```js
if ( activity.dependentOrigin?.active === false ) continue;
if ( !rollData && activity.uses?.recovery.some(u => u.type === "formula") ) rollData = this.parent.getRollData();
```

Fyorl commented on module/documents/actor/actor.mjs:
```diff
@@ -2571,17 +2571,20 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
+    if ( recoverShortRestUses ) recovery.unshift(["sr", 1]);
+    if ( recoverLongRestUses ) recovery.unshift(["lr", 1]);
+    if ( (recoverDailyUses || config.newDay) && game.settings.get("dnd5e", "calendarConfig").manualRecovery ) {
+      const days = (game.settings.get("dnd5e", "restVariant") === "gritty") && (config.type === "long") ? 7 : 1;
```
```suggestion
    if ( (recoverDailyUses || config.newDay) && dnd5e.settings.calendarConfig.manualRecovery ) {
      const days = (dnd5e.settings.restVariant === "gritty") && (config.type === "long") ? 7 : 1;
```

arbron committed:
[#2291, #5770] Swap to cached settings, optimize, fix JSDoc
